prov/util/src/util_attr.c

```diff
@@ -56,11 +57,167 @@ static int fi_valid_addr_format(uint32_t prov_format, uint32_t user_format)
+		if (!tok[i])
+			break;
+	}
+	if (i < max_tok)
+		goto err;
+
+	if (!get_prefix) {
+		/* Handle the case where there is a delim in token itself */
+		for (i = 0; i < max_tok; i++)
+			parsed_len += strlen(tok[i]) + 1;
+		parsed_len--;
+
+		if (strlen(name) > parsed_len)
+			tok[max_tok - 1][strlen(tok[max_tok - 1])] = *delim;
+	}
+
+	return 0;
+err:
+	free(name_dup);
+	return -FI_EOTHER;
+}
+
+int utilx_tr_layer_domain_attr(struct fi_domain_attr *layer_attr,
+		struct fi_domain_attr *base_attr)
+{
+	char *tok[2];
+	int ret;
+
+	ret = utilx_parse_name(layer_attr->name, 2, tok, 0);
+	if (ret)
+		return ret;
+	if (!(base_attr->name = strdup(tok[1])))
+		ret = -FI_ENOMEM;
+	free(tok[0]);
+	return ret;
+}
+
+int utilx_tr_layer_fabric_attr(struct fi_fabric_attr *layer_attr,
+		struct fi_fabric_attr *base_attr)
+{
+	char *tok[3];
+	int ret;
+
+	ret = utilx_parse_name(layer_attr->name, 3, tok, 0);
+	if (ret)
+		return ret;
+	if (!(base_attr->name = strdup(tok[2]))) {
+		ret = -FI_ENOMEM;
+		goto out;
+	}
+	if (!(base_attr->prov_name = strdup(tok[1])))
+		ret = -FI_ENOMEM;
+out:
+	free(tok[0]);
+	return ret;
+}
+
+int utilx_tr_base_domain_attr(struct fi_domain_attr *base_attr, char *prefix,
+		struct fi_domain_attr *layer_attr)
+{
+	size_t len;
+	len = strlen(prefix) + strlen(base_attr->name) + 2;
+	if (!(layer_attr->name = malloc(len)))
+		return -FI_ENOMEM;
+	snprintf(layer_attr->name, len, "%s_%s", prefix, base_attr->name);
+	return 0;
+}
+
+/* Construct fabric name for the layered provider in the following format:
+ * <layered_prov>_<base_prov_name>_<base_prov_fabric_name>
+ * e.g. rxm_verbs_IB-1234 */
+int utilx_tr_base_fabric_attr(struct fi_fabric_attr *base_attr, char *prefix,
+		struct fi_fabric_attr *layer_attr)
+{
+	size_t len;
+	len = strlen(prefix) + strlen(base_attr->name) +
+		strlen(base_attr->prov_name) + 3;
+	if (!(layer_attr->name = malloc(len)))
+		return -FI_ENOMEM;
+	snprintf(layer_attr->name, len, "%s_%s_%s", prefix, base_attr->prov_name,
+			base_attr->name);
+	return 0;
+}
+
+int utilx_getinfo(uint32_t version, const char *node, const char *service,
+			uint64_t flags, const struct fi_provider *prov,
+			const struct fi_info *prov_info, struct fi_info *hints,
+			utilx_tr_layer_info_t tr_layer_info,
+			utilx_tr_base_info_t tr_base_info,
+			int get_base_info, struct fi_info **info)
+{
+	struct fi_info *base_hints, *base_info;
+	int ret;
+
+	ret = fi_check_info(prov, prov_info, hints, FI_CHECK_LAYERED);
+	if (ret)
+		return ret;
+
+	if (!(base_hints = tr_layer_info(hints)))
+		return -FI_ENOMEM;
+
+	ret = fi_getinfo(version, node, service, flags, base_hints, &base_info);
+	if (ret)
+		goto out;
+
+	if (get_base_info) {
+		*info = base_info;
+	} else {
+		if (!(*info = tr_base_info(base_info))) {
+			ret = -FI_ENOMEM;
+			goto out;
+		}
+		fi_freeinfo(base_info);
+	}
+out:
+	fi_freeinfo(base_hints);
+	return ret;
+}
+
+int fi_check_name(char *user_name, char *prov_name, enum fi_check_type type)
```

Comment:
From just the perspective of this call, the name check is whether the name is an exact match or a prefix match.  I would change the API to reflect this, rather than assume it based on the provider layering.